### PR TITLE
Remove some global imports from std.file

### PR DESCRIPTION
**Pull request by JackStouffer**

### CODE REVIEW

File: std/file.d [hunk 1]
```diff
@@ -2607,6 +2633,7 @@ else version (NetBSD)
     }
     else version (FreeBSD)
     {
+        import std.exception : errnoEnforce;
```
Comment (ZombineDev) [inline std/file.d]: You need to add `assumeUnique` to list of imports.

**Comment by ZombineDev** [comment]
You need to rebase, because one commit from a different PR has sneaked in. You also need to add missing import for FreeBSD in one place. Assuming the autotester is green after that, all LGTM.

**Comment by JackStouffer** [comment]
>  because one commit from a different PR has sneaked in

No, I just decided to split it up to make it easier to review.

> You also need to add missing import for FreeBSD in one place

Thanks

**Comment by JackStouffer** [comment]
Looks good to go

File: std/file.d [hunk 2]
```diff
@@ -59,6 +57,9 @@ else
 package @property string deleteme() @safe
 {
     import std.process : thisProcessID;
+    import std.conv : to;
+    import std.path : buildPath;
+
     static _deleteme = "deleteme.dmd.unittest.pid";
```
Comment (joakim-noah) [inline std/file.d]: Place in alphabetical order.

File: std/file.d [hunk 3]
```diff
@@ -59,6 +57,9 @@ else
 package @property string deleteme() @safe
 {
     import std.process : thisProcessID;
+    import std.conv : to;
+    import std.path : buildPath;
```
Comment (joakim-noah) [inline std/file.d]: Place in alphabetical order, ie before `std.process`.

File: std/file.d [hunk 4]
```diff
@@ -2474,6 +2482,8 @@ else version(Posix) string readLink(R)(R link)
 version(Posix) @safe unittest
 {
     import std.string;
+    import std.exception : assertThrown;
```
Comment (joakim-noah) [inline std/file.d]: Place before `std.string`.

File: std/file.d [hunk 5]
```diff
@@ -3482,6 +3503,7 @@ version(Windows) unittest
 version(Posix) unittest
 {
     import std.process : executeShell;
+    import std.exception : enforce, collectException;
```
Comment (joakim-noah) [inline std/file.d]: Move up one line.

File: std/file.d [hunk 6]
```diff
@@ -3296,7 +3321,7 @@ private void copyImpl(const(char)[] f, const(char)[] t, const(FSChar)* fromz, co
         //cenforce(core.sys.posix.sys.stat.fstat(fdr, &statbufr) == 0, f, fromz);
 
         immutable fdw = core.sys.posix.fcntl.open(toz,
-                O_CREAT | O_WRONLY, octal!666);
+                O_CREAT | O_WRONLY, std.conv.octal!666);
         cenforce(fdw != -1, t, toz);
```
Comment (joakim-noah) [inline std/file.d]: I don't see an `octal` defined anywhere else: can't you just turn the static import above into a selective one?

File: std/file.d [hunk 7]
```diff
@@ -3569,6 +3597,7 @@ private struct DirIteratorImpl
         bool toNext(bool fetch, WIN32_FIND_DATAW* findinfo)
         {
             import core.stdc.wchar_ : wcscmp;
+            import std.path : buildPath;
```
Comment (joakim-noah) [inline std/file.d]: This import doesn't appear to be used, remove.

File: std/file.d [hunk 8]
```diff
@@ -3831,9 +3860,15 @@ unittest
...
 {
-    import std.algorithm;
-    import std.range;
-    import std.process;
+    import std.algorithm.searching : startsWith;
+    import std.algorithm.comparison : equal;
+    import std.algorithm.iteration : map;
+    import std.range.primitives : walkLength;
+    import std.process : thisProcessID;
+    import std.path : dirEntries, buildPath, absolutePath;
```
Comment (joakim-noah) [inline std/file.d]: Remove unnecessary static import of `std.path.absolutePath` below, just `absolutePath` should do.

File: std/file.d [hunk 9]
```diff
@@ -2501,6 +2518,7 @@ version(Posix) unittest // input range of dchars
 version(Windows) string getcwd()
 {
     import std.utf : toUTF8;
+    import std.conv : to;
```
Comment (joakim-noah) [inline std/file.d]: Move up one line.

File: std/file.d [hunk 10]
```diff
@@ -2568,6 +2586,7 @@ else version (NetBSD)
 {
     version (OSX)
     {
+        import std.conv : to;
```
Comment (joakim-noah) [inline std/file.d]: Move down one line.

File: std/file.d [hunk 11]
```diff
@@ -3831,9 +3860,15 @@ unittest
...
-    import std.range;
-    import std.process;
+    import std.algorithm.searching : startsWith;
+    import std.algorithm.comparison : equal;
+    import std.algorithm.iteration : map;
+    import std.range.primitives : walkLength;
+    import std.process : thisProcessID;
+    import std.path : dirEntries, buildPath, absolutePath;
+    import std.conv : to;
+    import std.array : array;
```
Comment (joakim-noah) [inline std/file.d]: As long as you're greatly expanding this list of imports, please alphabetically order them.

**Comment by JackStouffer** [comment]
As I stated before, I don't agree with your alphabetical ordering. I believe it introduces unnecessary diffs into the commit that don't add anything.

Thanks for pointing out the other problems though

**Comment by joakim-noah** [comment]
I know your opinion, which is why I left the pre-existing unordered imports alone and only pointed out diffs where you were gratuitiously placing your new imports out of alphabetical order.  In all these instances, there is no reason not to put them in order instead, as you'd only be moving around your own new imports, not touching any existing imports.  There are no "unnecessary diffs" pointed out above, so there's no reason not to change them.

**Comment by JackStouffer** [comment]
@joakim-noah Fixed

File: std/file.d [hunk 12]
```diff
@@ -3285,7 +3329,7 @@ private void copyImpl(const(char)[] f, const(char)[] t, const(FSChar)* fromz, co
     else version(Posix)
     {
         import core.stdc.stdio;
-        static import std.conv;
+        import std.conv;
```
Comment (joakim-noah) [inline std/file.d]: Make these selective.

**Comment by JackStouffer** [comment]
@joakim-noah Fixed

**Comment by joakim-noah** [comment]
Thanks, LGTM.

**Comment by WalterBright** [comment]
I don't want to pull this with the travis-ci failing.

**Comment by joakim-noah** [comment]
The Travis failure seems to be a style issue in an unrelated module, std.internal.phobosinit, ie a problem in master.

**Comment by WalterBright** [comment]
The travis-cl is passing on other PRs. How about kicking it so the test is re-run for this PR?

**Comment by schveiguy** [comment]
The phobosinit issue that travis flags has been fixed in master. How do you get travis to rerun when the base branch has changed?

**Comment by JackStouffer** [comment]
@WalterBright @schveiguy Done

**Comment by schveiguy** [comment]
With the auto-tester, I can go in and flag a test as a frivolous failure to make it run again. Can you do something similar with travis?

**Comment by JackStouffer** [comment]
No, I just changed the wording of one of the commits to change the hash and forced pushed

**Comment by schveiguy** [comment]
@JackStouffer Did some digging, yes there is a rebuild button for group members, once I signed in from github on travis, I could see it. I will keep this in mind next time!

I also found out they purposely avoid rebuilding all PRs when the upstream branch changes to avoid building too much at once.

**Comment by JackStouffer** [comment]
@schveiguy @WalterBright passing now

**Comment by schveiguy** [comment]
I'm hesitant to pull. This seems a lot more verbose... Are we sure that this is better considering how the import system has been fixed?

**Comment by JackStouffer** [comment]
> I'm hesitant to pull. This seems a lot more verbose... Are we sure that this is better considering how the import system has been fixed?

While it's likely that std.conv has already been imported by another piece of code, the same cannot be said for std.exception or std.path. Moving these to the local scope reduces the likelihood that either of those modules need to be imported for a program.

**Comment by schveiguy** [comment]
> Moving these to the local scope reduces the likelihood that either of those modules need to be imported for a program.

Are you sure that's the case? If I import locally, then it still needs to be imported. It doesn't go into the larger namespace, but the import still needs to happen. This isn't necessarily true for templates, but many of these aren't templates.

**Comment by JackStouffer** [comment]
@schveiguy I was under the impression that an import in a function scope only occurs if that function is used in a program; this is what I gathered from one of Andrei's talks pitching D.

**Comment by schveiguy** [comment]
I used the -v switch, and yes, this does actually happen (the compiler only parses the local import if you call the function). So it is a good thing, thanks!

**Pull request merged**